_posts/2019-12-08-what-is-a-shell.md

```diff
@@ -0,0 +1,166 @@
+you may have to provide additional input to the command,
+which is called a **flag**.
+For example, `ls -l` provides a "long list" format which also shows file permissions, file creation and file access information.
+That particular flag will work in Bash and most other Bash inspired shells, but it doesn't work in Powershell.
+
+Powershell provides a similar features in the form of a cmdlet, pronounced _command-let_.
+You can try `ls` in Powershell and it will work, but also try `man ls` to get the manual page.
+You'll find that the cmdlet is actually named `Get-ChildItem`.
+If you use `alias ls` in Powershell,
+you will see the command is an alias for the Get-ChildItem.
+
+It's somewhat interesting that `ls` in Bash is similar to `ls -Name` in Powershell
+and `ls -l` in Bash is similar to the default `ls` in Powershell.
+
+There are plenty of other executable programs that operating systems have to offer.
+For Linux/MacOS,
+you most likely have the path `/usr/bin` in your `PATH` variable,
+which is what directories your shell will look in to find executable programs.
+Try `ls -l /usr/bin` to see a list of executables which you can run from the command line.
+If you're curious about any of them,
+all you need is the manual page by typing the executable name preceded by `man`.
+For example `man zip`.
+Similar functionality can be achieved in Powershell by using `Get-Command`.
+
+### Manipulating Data
+
+In the previous section we saw how a user may list the available commands or executables they can run via their shell.
+If you've been trying these on your own system,
+you may have noticed an overwhelming amount of output.
+Sometimes you may need all of it,
+but often you might want a subset of the data.
+
+We'll continue with `ls`,
+as it is fairly simple to work with.
+In Bash, if you want to `ls` only directories,
+or in other words, _filter entries by the type directory_,
+then you may use the command `ls -d */`.
+The logic behind this is that the `-d` flag will output the / character after a directory and then we're using a glob to match anything that ends with the /.
+If you're in powershell, you can take advantage of the `-Directory` flag.
+The full command will look like `ls -Directory`.
+
+These shells differ in their approach to how data is handled.
+Bash handles data like raw text while Powershell seems to handle the data as objects.
+Both approaches have advantages and disadvantages,
+however if you need to context switch between shells this can be a lot of mental overhead.
+
+Sometimes simple data manipulation goes well beyond one command.
+If you are experienced in your shell of choice,
+you've probably _piped_ commands together or even written scripts.
+The idea is that commands can take some input, maybe from the user, maybe from something more formalized like standard-in,
+and then pass it into a program.
+Shell programs often output data via standard-out allowing you to continue this pattern of linking programs together.
+
```

Comment:
Maybe in this section you want to give a brief introduction to what you can do with piping? Maybe something simply like getting the last 5 lines of a file `cat <file> | tail -n5`